### ISSUE TITLE
How to optimize `GradleRunner` performance? Adding more cores makes it slower 🤔

### ISSUE THREAD

**Issue by SUPERCILEX** [gradle/gradle#13421]
This isn't a bug persay, but I'm baffled by this behavior. I have have a project which [runs tests in parallel](https://github.com/Triple-T/gradle-play-publisher/blob/367550b9674c3a0f6f8ec7d8df38c006152b4f36/build.gradle.kts#L102-L104), but I've noticed that the more cores I add (whether through JUnit's executor or the `maxParallelForks` param), my build times get slower. I'm currently [manually locking](https://github.com/Triple-T/gradle-play-publisher/blob/367550b9674c3a0f6f8ec7d8df38c006152b4f36/play/plugin/src/test/kotlin/com/github/triplet/gradle/play/helpers/IntegrationTestBase.kt#L54) the concurrency on the `GradleRunner` which yields a ~3 min build speed improvement over using all 12 of my cores. I've checked and memory isn't the issue. Is there something I'm not understanding about the testkit daemon? Even running two tests at once is slower. Shouldn't two daemons be faster than one at scale since the cost of starting them up is only a few seconds?

**Comment by runningcode** [comment]
Hey can you tell us more about how you verified that memory is not an issue?
Based on your max parallel forks parameters and the number of cores on your machine, you would likely need 12 cores * 2GB per runner + 2GB for the standard daemon. How much memory does this machine have?

**Comment by SUPERCILEX** [comment]
To clarify, I tested max parallel forks and the jupiter executor separately. So in the max parallel forks scenario, I stuck with the default which I believe is 512m (or 6GB for 12 cores) which is fine on a 16GB machine. And ignoring maxed out cores for a second, it was still slower with `maxParallelForks = 2`. In the jupiter executor scenario, I have 2GB since there should only be a single fork running.

Here's a build scan with jupiter running on all cores, but with manual locking so only one GradleRunner can execute at a time: https://scans.gradle.com/s/5pxlfqa5iukde/tests#:play:plugin. Now here's the exact same scenario WITHOUT locking: https://scans.gradle.com/s/dhns4a2aylhke/tests#:play:plugin. It's almost 3X slower!!! I have some screenshots below of memory and CPU usage for the test without locking (that should in theory be faster). Memory peaked at 11/16GB, and interestingly, CPU maxed out my cores at the beginning of the test, but then only hovered around 10-20% utilization for the rest of the build. This isn't thermal throttling because I monitored my clock speeds and they stayed consistently around ~3.5 Ghz (which was the same as the build with locking). 

The CPU graph is why I'm thinking there's file contention of some sort, or perhaps the OS is wasting cycles switching between the Gradle daemons? I'm not sure how to investigate this further.

![image](https://user-images.githubusercontent.com/9490724/84741766-65ef6680-af64-11ea-91e8-0cf9e97a3c80.png)
![image](https://user-images.githubusercontent.com/9490724/84741760-62f47600-af64-11ea-8f34-a2d2e6bb7420.png)
